NETProvider/src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/SrpClient.cs

```diff
@@ -0,0 +1,239 @@
+ *	   language governing rights and limitations under the License.
+ *
+ *	Copyright (c) 2015 Hajime Nakagami<nakagami@gmail.com>
+ *	All Rights Reserved.
+ *
+ */
+
+// http://srp.stanford.edu/design.html
+// https://github.com/FirebirdSQL/jaybird/blob/master/src/main/org/firebirdsql/gds/ng/wire/SrpClient.java
+
+using System;
+using System.Linq;
+using System.Collections.Generic;
+using System.Numerics;
+using System.Security.Cryptography;
+using System.Globalization;
+
+namespace FirebirdSql.Data.Client.Managed.Version10
+{
+	public class SrpClient {
+		private const int SRP_KEY_SIZE = 128;
+		private const int SRP_SALT_SIZE = 32;
+		private static BigInteger N = BigInteger.Parse("00E67D2E994B2F900C3F41F08F5BB2627ED0D49EE1FE767A52EFCD565CD6E768812C3E1E9CE8F0A8BEA6CB13CD29DDEBF7A96D4A93B55D488DF099A15C89DCB0640738EB2CBDD9A8F7BAB561AB1B0DC1C6CDABF303264A08D1BCA932D1F1EE428B619D970F342ABA9A65793B8B2F041AE5364350C16F735F56ECBCA87BD57B29E7", NumberStyles.HexNumber);
+		private static BigInteger g = new BigInteger(2);
+		private static BigInteger k = BigInteger.Parse("1277432915985975349439481660349303019122249719989");
+
+		private static RandomNumberGenerator random = RandomNumberGenerator.Create();
+		private static byte[] SEPARATOR_BYTES = System.Text.Encoding.UTF8.GetBytes(":");
+
+		private BigInteger publicKey;	/* A */
+		private BigInteger privateKey;	/* a */
+		private byte[] proof;			/* M */
+		private byte[] sessionKey;		/* K */
+
+		private static BigInteger fromBigByteArray(byte [] b) {
+			Array.Reverse(b);
+			return new BigInteger(b.Concat(new byte[] { 0 }).ToArray());
+		}
+
+		private static byte[] toBigByteArray(BigInteger n) {
+			byte[] b = n.ToByteArray();
+			int length = b.Length;
+			if (b[length-1] == 0) {
+				length -= 1;
+			}
+			byte[] b2 = new byte[length];
+			Array.Copy(b, 0, b2, 0, length);
+			Array.Reverse(b2);
+			return b2;
+		}
+
+		private static String toHexString(byte[] b) {
+			return BitConverter.ToString(b).Replace("-", string.Empty);
+		}
+
+		private static byte[] fromHexString(String s) {
+			byte[] b = new byte[s.Length / 2];
+			for (int i = 0; i < b.Length; i++ ) {
+				b[i] = Convert.ToByte(s.Substring(i*2, 2), 16);
+			}
+
+			return b;
+		}
+
+		private static byte[] sha1(params byte[][] ba) {
+			List<byte> mergedList = new List<byte>(ba.Sum(b =>b.Length));
+			SHA1 hash = SHA1.Create();
+			foreach(byte[] b in ba) {
+				mergedList.AddRange(b);
+			}
+			return hash.ComputeHash(mergedList.ToArray());
+		}
+
+		private static byte[] pad(BigInteger n) {
+			byte[] bn = toBigByteArray(n);
+			if (bn.Length > SRP_KEY_SIZE) {
+				byte[] buf = new byte[SRP_KEY_SIZE];
+				Array.Copy(bn, bn.Length - SRP_KEY_SIZE, buf, 0, SRP_KEY_SIZE);
+				return buf;
+			}
+			return bn;
+		}
+
+		private static BigInteger getScramble(BigInteger x, BigInteger y) {
+			return fromBigByteArray(sha1(pad(x), pad(y)));
+		}
+
+		private BigInteger getSecret() {
```

Comment:
Is this method used in a single thread scenarios?
If it can be used in multithreaded scenarios, ```random``` field should be removed and ```RandomNumberGenerator``` instances created in method body.
Same issue with ```getSalt```